docs: add docs on using custom entities with DB-less kong (#652)
Kong/kubernetes-ingress-controller: docs/guides/configuring-custom-entities.md

which are used to store additional configuration.
This guide elaborates on how such custom entities can be used Kong Ingress
Controller.

> Note: All entities shipped with Kong are supported by Kong Ingress Controller
out of the box. This guide applies only if you have a custom entity in your
plugin. Custom plugins have first-class support in Kong Ingress Controller
via the `KongPlugin` CRD. Please read [this](../setting-up-custom-plugins.md)
guide for details.

## Caveats

- The feature discussed in this guide apply for DB-less deployments of Kong.
  The feature is not supported for deployments where Kong is used with a
  database or Kong is used in hybrid mode.
  For these deployments, it is recommended to configure custom entities
  directly using Kong's Admin API.
- Custom entities which have a foreign relation with other core entities in Kong
  are not supported. Only entities which can exist by themselves and then
  be referenced via plugin configuration are supported.

## Creating a JSON representation of the custom entity

In this section, we will learn how to create a JSON representation of
a custom entity.

> Suppose you have a custom entity in Kong with the following schema in Kong:

Reviewer: ```suggestion
Suppose you have a custom entity with the following schema in your plugin source:
```
"schema in Kong" also works, though is a bit less specific. Main thing is to not have "in Kong" repeated twice.

```lua 
{
  name                = "xkcds",
  primary_key         = { "id" },
  cache_key           = { "name" },
  endpoint_key        = "name",
  fields = {
    { id = typedefs.uuid },
    {
      name = {
        type= "string",
        required = true,
        unique = true,
      },
    },
    {
      url = {
        type = "string",
        required = true,
      },
    },
    { created_at = typedefs.auto_timestamp_s },
    { updated_at = typedefs.auto_timestamp_s },
  },
}
```

> An instance of such an entity would like:

Reviewer: ```suggestion
An instance of such an entity would look like:
```

```json
{
  "id": "385def6e-3059-4929-bb12-d205e97284c5",
  "name": "Bobby Drop Tables",
  "url": "https://xkcd.com/327/"
}
```

Multiple instances of such an entity are represented as follows:

```json
{
  "xkcds": [
    {

>       "id": "385def6e-3059-4929-bb12-d205e97284c5",

Reviewer: Are the IDs necessary? They'll be present if this configuration is exported from an existing Kong instance, but not necessarily if they're generated by some other system. If they can be omitted, we should call out that they're unnecessary/will be added automatically when they're stored in Kong.

Author: > If they can be omitted, we should call out that they're unnecessary/will be added automatically when they're stored in Kong.

It is likely that the plugin configuration will refer to these custom entities via IDs. To keep configuration declarative, it is important to specify ID as we can't rely on autogen ID.

      "name": "bobby_tables",
      "url": "https://xkcd.com/327/"
    },
    {
      "id": "d079a632-ac8d-4a9a-860c-71de82e8fc11",
      "name": "compiling",
      "url": "https://xkcd.com/303/"
    }
  ]
}
```

If you have more than one custom entities that you would like to configure
then you can create other entities by specifying the entity name at the root

> level of the JSON and then list the entities as a JSON array.

Reviewer: This seems to imply that you can provide configuration without using entity names mapped to arrays of that entities--can you do so? I don't see a way it's possible, since there's no other way to indicate what type of entity exists in the secret.

In either case, I'd recommend including a second entity type in the example to show the format in practice. The schema for the additional entity should be omitted for brevity.

Author: I reworded a little bit to avoid adding another entity (this doc is already pretty hard to grok).

To configure custom entities in a DB-less instance of Kong,
you first need to create such a JSON representation of your entities.

## Configuring the custom entity secret

Once you have the JSON representation, we need to store the configuration
inside a Kubernetes Secret.
The following command assumes the filename to be `entities.json` but you can
use any other filename as well:

```bash
$ kubectl create secret generic -n kong kong-custom-entities --from-file=config=entities.json
secret/kong-custom-entities created
```

Some things to note:

> - The key inside the secret should be `config`. This is not configurable at the

Reviewer: ```suggestion
- The key inside the secret must be `config`. This is not configurable at the
```

  moment.

> - The secret should be accessible by the Ingress Controller. The recommended

Reviewer: ```suggestion
- The secret must be accessible by the Ingress Controller. The recommended
```

  is running.
 
## Configure the Ingress Controller

Once you have the secret containing the custom entities configured,
you need to instruct the controller to read the secret and sync the custom
entities to Kong.

To do this, you need to add the following environment variable to the
`ingress-ccontroller` container:

```yaml
env:
- name: CONTROLLER_KONG_CUSTOM_ENTITIES_SECRET
  value: kong/kong-custom-entities
```

This value of the environment variable takes the form of `<namespace>/<name>`.
You need to configure this only once.

This instructs the controller to watch the above secret and configure Kong
with any custom entities present inside the secret.
If you change the configuration and update the secret with different entities,
the controller will dynamically fetch the updated secret and configure Kong.

## Verification

You can verify that the custom entity was actually created in Kong's memory
using the `GET /xkcds` (endpoint will differ based on the name of the entity)
on Kong's Admin API.
For most deployments, you will need to exec inside the Kong container and
perform:
```bash
$ curl -k https://localhost:8444/<entity-name>

> ```

Reviewer: We no longer package curl with our images, and will often have no non-local listen (so spawning a test pod to run curl isn't an option:
> You can forward traffic from your local machine to the Kong Pod to access it:
```
$ kubectl port-forward kong/kong-pod-name 8444:8444
```
> and in a separate terminal:
```
 $ curl -k https://localhost:8444/<entity-name>
```
